TestingLowerBounds/ForMathlib/LeftRightDeriv.lean

```diff
@@ -0,0 +1,190 @@
+
+lemma rightDeriv_def (f : ℝ → ℝ) (x : ℝ) : rightDeriv f x = derivWithin f (Ioi x) x := rfl
+
+noncomputable
+def leftDeriv (f : ℝ → ℝ) : ℝ → ℝ := fun x ↦ derivWithin f (Iio x) x
+
+lemma leftDeriv_def (f : ℝ → ℝ) (x : ℝ) : leftDeriv f x = derivWithin f (Iio x) x := rfl
+
+lemma rightDeriv_eq_leftDeriv_apply (f : ℝ → ℝ) (x : ℝ) :
+    rightDeriv f x = - leftDeriv (f ∘ Neg.neg) (-x) := by
+  have h_map : MapsTo Neg.neg (Iio (-x)) (Ioi x) := fun _ (h : _ < -x) ↦ lt_neg_of_lt_neg h
+  have h_map' : MapsTo Neg.neg (Ioi x) (Iio (-x)) := fun _ (h : x < _) ↦ neg_lt_neg h
+  by_cases hf_diff : DifferentiableWithinAt ℝ f (Ioi x) x
+  swap
+  · rw [rightDeriv_def, leftDeriv_def, derivWithin_zero_of_not_differentiableWithinAt hf_diff,
+      derivWithin_zero_of_not_differentiableWithinAt, neg_zero]
+    contrapose! hf_diff
+    convert DifferentiableWithinAt.comp x hf_diff ((differentiable_neg _).differentiableWithinAt)
+      h_map' using 1
+    simp [Function.comp.assoc]
+  simp_rw [leftDeriv]
+  rw [derivWithin.comp _ ((neg_neg x).symm ▸ hf_diff) (differentiable_neg _).differentiableWithinAt
+    h_map (uniqueDiffWithinAt_Iio (-x)), neg_neg, ← rightDeriv_def, derivWithin_neg]
+  swap; · exact uniqueDiffWithinAt_Iio _
+  simp only [mul_neg, mul_one, neg_neg]
+
+lemma rightDeriv_eq_leftDeriv (f : ℝ → ℝ) : rightDeriv f = - leftDeriv (f ∘ Neg.neg) ∘ Neg.neg := by
+  ext x
+  simp [rightDeriv_eq_leftDeriv_apply]
+
+lemma leftDeriv_eq_rightDeriv_apply (f : ℝ → ℝ) (x : ℝ) :
+    leftDeriv f x = - rightDeriv (f ∘ Neg.neg) (-x) := by
+  simp [rightDeriv_eq_leftDeriv_apply, Function.comp.assoc]
+
+lemma leftDeriv_eq_rightDeriv (f : ℝ → ℝ) : leftDeriv f = - rightDeriv (f ∘ Neg.neg) ∘ Neg.neg := by
+  ext x
+  simp [leftDeriv_eq_rightDeriv_apply]
+
+namespace ConvexOn
+
+section Slope
+
+variable {𝕜 : Type*} [LinearOrderedField 𝕜] {s : Set 𝕜} {f : 𝕜 → 𝕜} {x : 𝕜}
+
+--This has already been merged in Mathlib, see #14015. When we bump remove this
+lemma slope_mono (hfc : ConvexOn 𝕜 s f) (hx : x ∈ s) : MonotoneOn (slope f x) (s \ {x}) :=
+  (slope_fun_def_field f _).symm ▸ fun _ hy _ hz hz' ↦ hfc.secant_mono hx (mem_of_mem_diff hy)
+    (mem_of_mem_diff hz) (not_mem_of_mem_diff hy :) (not_mem_of_mem_diff hz :) hz'
+
+lemma bddBelow_slope_Ioi_of_convexOn {f : ℝ → ℝ} (hfc : ConvexOn ℝ univ f) (x : ℝ) :
+    BddBelow (slope f x '' Ioi x) := by
+  refine bddBelow_iff_subset_Ici.mpr ⟨(slope f x (x - 1)), fun y ⟨z, (hz : x < z), hz'⟩ ↦ ?_⟩
+  simp_rw [mem_Ici, ← hz']
+  exact slope_mono hfc trivial (by simp) ⟨trivial, hz.ne'⟩ (by linarith)
+
+lemma bddAbove_slope_Iio_of_convexOn {f : ℝ → ℝ} (hfc : ConvexOn ℝ univ f) (x : ℝ) :
+    BddAbove (slope f x '' Iio x) := by
+  refine bddAbove_iff_subset_Iic.mpr ⟨(slope f x (x + 1)), fun y ⟨z, (hz : z < x), hz'⟩ ↦ ?_⟩
+  simp_rw [mem_Iic, ← hz']
+  exact slope_mono hfc (mem_univ x) ⟨trivial, hz.ne⟩ (by simp) (by linarith)
+
+end Slope
+
+variable {f : ℝ → ℝ} (hfc : ConvexOn ℝ univ f)
```

Comment:
Please don't put explicit hypotheses in `variable` lines. It makes the code hard to read. `(hfc : ConvexOn ℝ univ f)` should be moved to the lemma statements.